src/core/service.c

```diff
@@ -2644,6 +2644,10 @@ static void service_sigchld_event(Unit *u, pid_t pid, int code, int status) {
         else
                 assert_not_reached("Unknown code");
 
+        if (f != SERVICE_SUCCESS && status == SIGTERM &&
+            streq(basename(UNIT(s)->fragment_path), SPECIAL_INITRD_SWITCH_ROOT_SERVICE))
```

Comment:
If unit doesn't have the unit file, e.g. transient service, then NULL will be passed into `basename()` and that will trigger segmentation fault.

Reply:
Please replace `basename()` and `streq()` with `unit_has_name()` to check the name of the unit.

Reply:
> If unit doesn't have the unit file, e.g. transient service, then NULL will be passed into `basename()` and that will trigger segmentation fault.

According to the manual, segfaults should not occur:
If path is a NULL pointer or points to an empty string, then both dirname() and basename() return the string ".".
From: https://linux.die.net/man/3/basename

unit_has_name() is indeed more beautiful, we will modify it according to your suggestions.
Thanks.